lib/ruby-enum/enum/ecase.rb

```diff
@@ -0,0 +1,72 @@
+# frozen_string_literal: true
+
+module Ruby
+  module Enum
+    ##
+    # Adds a method to an enum class that allows for exhaustive matching on a value.
+    #
+    # @example
+    # class Color
+    #   include Ruby::Enum
+    #   include Ruby::Enum::Ecase
+    #
+    #   define :RED, :red
+    #   define :GREEN, :green
+    #   define :BLUE, :blue
+    # end
+    #
+    # Color.ecase(Color::RED, {
+    #   [Color::RED, Color::GREEN] => -> { puts "red or green" },
+    #   Color::BLUE => -> { puts "blue" },
+    # })
+    #
+    # Reserves the :else key for a default case:
+    # Color.ecase(Color::RED, {
+    #   [Color::RED, Color::GREEN] => -> { puts "red or green" },
+    #   else: -> { puts "blue" },
+    # })
+    module Ecase
+      def self.included(klass)
+        klass.extend(ClassMethods)
+      end
+
+      ##
+      # @see Ruby::Enum::Ecase
+      module ClassMethods
+        class ValuesNotDefinedError < StandardError
+        end
+
+        class NotAllCasesHandledError < StandardError
+        end
+
+        def ecase(value, cases)
+          validate_cases(cases)
+
+          filtered_cases = cases.select do |values, _block|
+            values = [values] unless values.is_a?(Array)
+            values.include?(value)
+          end
+
+          return cases[:else]&.call if filtered_cases.none?
+
+          results = filtered_cases.map { |_values, block| block.call }
+
+          # Return the first result if there is only one result
+          results.size == 1 ? results.first : results
+        end
+
+        private
+
+        def validate_cases(cases)
```

Comment:
Rename to `validate_cases!` since it raises?

Reply:
I think the convention is that you should only provide a method with a `!` if you also provide a safe method without it, which we don't do here. Also, I think this only applies to public methods. So I would leave it as it is if that's ok with you.